### ISSUE TITLE
Site creation after free checkout silently falls back to Action Scheduler (loopback fast-path nonce failure)

### ISSUE THREAD

**Issue by superdav42** [Ultimate-Multisite/ultimate-multisite#1231]
## Summary

`Membership::publish_pending_site_async()` has a loopback HTTP "fast path" to `admin-ajax.php?action=wu_publish_pending_site` that is supposed to start site creation immediately after checkout. On stacks where the calling request has just auto-logged-in the customer (typical free-checkout path), this loopback is silently rejected by `check_ajax_referer()` and the site is only created later by the Action Scheduler fallback (`wu_async_publish_pending_site`). On servers where the network main site's WP-cron ticks infrequently (e.g. once an hour), customers wait up to an hour for their site instead of seeing it immediately.

There is **no log entry** when this happens, so operators have no signal that the fast path is broken.

## Reproduction (production evidence)

- WP 7.0-RC2, Ultimate Multisite 2.5.x, Action Scheduler 3.9.3
- FrankenPHP (no `fastcgi_finish_request`), free `plan_id=8` checkout via the `free` gateway
- Membership 54 / customer 54 / user 1037 / blog 80 (`davidnewtest.sdaiagent.com`)

Timeline:

| GMT | Event |
|---|---|
| 16:53:52 | Membership 54 created (`wp_wu_memberships.date_created`) |
| 16:53:52 | Action 82464 `wu_async_publish_pending_site` enqueued with `scheduled_date_gmt = now` |
| 16:53:55 | `checkout_completed` event written |
| 17:00:03 | Blog 80 row created in `wp_blogs` (= when AS queue tick fired) |
| 17:00:36 | Action 82464 marked `complete` |

**~7-minute gap between checkout and site existence**, attributable entirely to waiting for the next AS queue tick. The loopback fast path never created the site.

Direct probe of the AJAX endpoint from the same host (no cookies):

```
$ curl -sS -o - -w 'HTTP %{http_code}\n' \
    'https://mygratis.site/wp/wp-admin/admin-ajax.php?action=wu_publish_pending_site&membership_id=54'
0HTTP 400
```

`check_ajax_referer()` rejected the request and emitted `0` + `die()` — exactly what the loopback fast-path receives.

## Root cause

`inc/models/class-membership.php:2110-2150`:

```php
$rest_path = add_query_arg([
    'action'        => 'wu_publish_pending_site',
    '_ajax_nonce'   => wp_create_nonce('wu_publish_pending_site'),
    'membership_id' => $this->get_id(),
], admin_url('admin-ajax.php'));

$request_args = [
    'cookies'   => wp_unslash($_COOKIE), // Cookies are needed for nonce check to work.
    'timeout'   => 10,
    'sslverify' => apply_filters('https_local_ssl_verify', false),
    'headers'   => $headers,
];

if ( ! function_exists('fastcgi_finish_request')) {
    $request_args['blocking'] = false;
}
$result = wp_remote_request($rest_path, $request_args);

if (is_wp_error($result)) {
    wu_log_add("membership-{$this->get_id()}", sprintf( /* … */ ));
}

wu_enqueue_async_action('wu_async_publish_pending_site', /* … */);
```

Three compounding defects:

1. **Stale cookies for new sessions.** `$_COOKIE` holds the cookies present in the *inbound* checkout request, before the customer was auto-logged-in. The calling PHP process now has a logged-in user (`wp_get_current_user()` returns the new customer), so `wp_create_nonce('wu_publish_pending_site')` is generated against that new session token. But the loopback receiver only sees the stale pre-login cookies, decodes a different (or empty) session token, and `wp_verify_nonce()` fails. Result: HTTP 400 / body `0`.

2. **Silent failure.** The caller only logs when `wp_remote_request` returns a `WP_Error`. A `400` response body `0` (the canonical AS nonce-failure signature) is just a successful HTTP transaction with the wrong body. No log file is written, no admin notice, nothing.

3. **`fastcgi_finish_request` polyfill detection is host-dependent.** Under FrankenPHP the function does not exist, so the caller sets `blocking => false`. That hides any HTTP-level signal anyway, because the caller never reads the response — even if we wanted to surface the `400`, we couldn't from this side.

When (1) trips, the fallback Action Scheduler entry is the only thing that publishes the site, which means site-creation latency becomes equal to "time until the next AS queue tick on the network main site." On installs with infrequent WP-cron pacing that latency can be tens of minutes.

## Proposed fix

Replace the cookie-based nonce with a server-side HMAC token that the calling process can mint without relying on `$_COOKIE` round-tripping, and surface non-2xx responses to the log.

**File to modify:** `inc/models/class-membership.php` (`publish_pending_site_async()` around line 2100)
**File to modify:** `inc/managers/class-membership-manager.php` (`publish_pending_site()` around line 111)

Suggested approach:

1. Introduce a short-lived (≤60s) HMAC token keyed on `wp_salt('auth') . $membership_id . $expires` minted by the caller:
   ```php
   $expires = time() + 60;
   $token   = hash_hmac('sha256', $this->get_id() . '|' . $expires, wp_salt('auth'));
   $rest_path = add_query_arg([
       'action'        => 'wu_publish_pending_site',
       'membership_id' => $this->get_id(),
       'wu_token'      => $token,
       'wu_expires'    => $expires,
   ], admin_url('admin-ajax.php'));
   ```
2. In `Membership_Manager::publish_pending_site()` replace `check_ajax_referer()` with:
   - If `wu_token` + `wu_expires` are present and verify with `hash_equals()` against the same HMAC, proceed.
   - Otherwise fall back to existing `check_ajax_referer('wu_publish_pending_site')` so the admin-modal/manual flow keeps working.
3. After `wp_remote_request()`, also log non-2xx responses (use `wp_remote_retrieve_response_code()`), so future loopback failures surface immediately:
   ```php
   if (is_wp_error($result)) {
       wu_log_add("membership-{$this->get_id()}", /* current message */);
   } else {
       $code = (int) wp_remote_retrieve_response_code($result);
       if ($code < 200 || $code >= 300) {
           wu_log_add(
               "membership-{$this->get_id()}",
               sprintf('Loopback fast-path returned HTTP %d — falling back to Action Scheduler.', $code)
           );
       }
   }
   ```
4. Drop the `'cookies' => wp_unslash($_COOKIE)` line and the `headers['Authorization'] = 'Basic …'` block — both become unnecessary once the token is in the URL.

This preserves the existing AS fallback and the admin-modal-driven `publish_pending_site` path, but makes the auto-publish-after-checkout loopback robust against the stale-cookie case and visible when it fails.

## Verification

- `tests/WP_Ultimo/Managers/Membership_Manager_Test.php`:
  - Add a test that calls `publish_pending_site()` AJAX action with a valid HMAC token (no cookies, no nonce) and asserts the membership's pending site is published.
  - Add a test that asserts a stale/forged token is rejected with HTTP 400.
  - Add a test that asserts the existing nonce path (admin "Publish pending site" modal) still works.
- Manual reproduction on a FrankenPHP install:
  1. With current `master`: complete a free checkout, observe `wp_blogs` row appears only on next AS queue tick.
  2. With patch: complete a free checkout, observe `wp_blogs` row appears within ≤2s of `checkout_completed` event.
- Log inspection: with a deliberately invalid token, confirm `wu-logs/membership-{ID}.log` contains an HTTP-code line.

## Out of scope

- Why AS queue runner only ticks hourly on a particular server — that is a stack-side WP-cron pacing issue, not a plugin bug. Tracked separately by the reporter.
- The `force_publish_sites_sync` setting already provides a workaround (sync site creation inside the checkout request); this brief targets the async path that is supposed to be the default.

## Files for the worker

- `inc/models/class-membership.php` — `publish_pending_site_async()` (~line 2100)
- `inc/managers/class-membership-manager.php` — `publish_pending_site()` (~line 111), `async_publish_pending_site()` (~line 145)
- `tests/WP_Ultimo/Managers/Membership_Manager_Test.php` — extend nonce/token assertions

---
[aidevops.sh](https://aidevops.sh) v3.17.8 plugin for [OpenCode](https://opencode.ai) v1.15.5 with claude-sonnet-4-20250514 spent 3h 27m on this as a headless worker.

**Comment by superdav42** [comment]
DISPATCH_CLAIM nonce=d641e00f9aaa134b08e3b3a8f9af85ab runner=superdav42 ts=2026-05-20T20:47:01Z max_age_s=1800 version=3.17.9 opencode_version=1.15.5

**Comment by superdav42** [comment]
Dispatching worker (deterministic).
- **Worker PID**: 1613677
- **Model**: openai/gpt-5
- **Tier**: standard
- **Runner**: superdav42
- **aidevops**: v3.17.9
- **OpenCode**: v1.15.5
- **Issue**: #1231

**Comment by superdav42** [comment]
CLAIM_RELEASED reason=worker_failed runner=superdav42 ts=2026-05-20T20:48:17Z aidevops_version=3.17.9 opencode_version=1.15.5

**Comment by superdav42** [comment]
DISPATCH_CLAIM nonce=62c70b17a20f2996157c00a141322975 runner=superdav42 ts=2026-05-20T23:59:03Z max_age_s=1800 version=3.17.14 opencode_version=1.15.6

**Comment by superdav42** [comment]
Dispatching worker (deterministic).
- **Worker PID**: 3661331
- **Model**: openai/gpt-5
- **Tier**: standard
- **Runner**: superdav42
- **aidevops**: v3.17.14
- **OpenCode**: v1.15.6
- **Issue**: #1231

**Comment by superdav42** [comment]
CLAIM_RELEASED reason=worker_failed runner=superdav42 ts=2026-05-21T00:00:12Z aidevops_version=3.17.14 opencode_version=1.15.6

**Comment by superdav42** [comment]
DISPATCH_CLAIM nonce=6bb7b3b3a37523a5341169bb4eea17bb runner=superdav42 ts=2026-05-21T00:03:42Z max_age_s=1800 version=3.17.14 opencode_version=1.15.6

**Comment by superdav42** [comment]
Dispatching worker (deterministic).
- **Worker PID**: 3705077
- **Model**: openai/gpt-5
- **Tier**: standard
- **Runner**: superdav42
- **aidevops**: v3.17.14
- **OpenCode**: v1.15.6
- **Issue**: #1231

**Comment by superdav42** [comment]
CLAIM_RELEASED reason=worker_failed runner=superdav42 ts=2026-05-21T00:04:46Z aidevops_version=3.17.14 opencode_version=1.15.6

**Comment by superdav42** [comment]
DISPATCH_CLAIM nonce=6841ed6211b83d014f34dff76e633788 runner=superdav42 ts=2026-05-21T00:07:47Z max_age_s=1800 version=3.17.14 opencode_version=1.15.6

**Comment by superdav42** [comment]
Dispatching worker (deterministic).
- **Worker PID**: 3742401
- **Model**: anthropic/claude-haiku-4-5
- **Tier**: standard
- **Runner**: superdav42
- **aidevops**: v3.17.14
- **OpenCode**: v1.15.6
- **Issue**: #1231

**Comment by superdav42** [comment]
CLAIM_RELEASED reason=worker_complete runner=superdav42 ts=2026-05-21T00:36:57Z aidevops_version=3.17.14 opencode_version=1.15.6

**Comment by superdav42** [comment]
Completed via PR #1233.
- **Attempts**: 1
- **Duration**: ~29 minutes (dispatched 00:07, merged 00:36 UTC)
- **Fix**: HMAC token-based loopback verification for site creation (replaces nonce/cookie-based fast-path)



---
[aidevops.sh](https://aidevops.sh) v3.17.18 plugin for [OpenCode](https://opencode.ai) v1.15.6 with claude-sonnet-4-6 spent 11m and 18,199 tokens on this as a headless worker. Overall, 7h 39m since this issue was created.
